Integrating Websockets in Directus 🕸️🧦

br41nslug opened the pull request:
Integrating two-way communication into the Directus server to enable real-time functionality.

> **Warning**
> Note: this description is probably a bit outdated by now, please use the docs as reference.

## MVP Features

- A CRUD implementation over WebSockets
- A REST Subscriptions implementation
- GraphQL Subscriptions over WebSockets
- Three authentication modes: `public`, `handshake`, `strict`
- Authentication refresh with an open socket
- Heartbeat signal to keep the connection alive
- Follows the Directus permission model
- Message rate limiting and connection limiting
- Horizontal scaling with the Messenger
- Extensible event driven design

## Configuration

### Environment Variables

| Variable           | Description                                | Type    | Default Value |
|--------------------|--------------------------------------------|---------|---------------|
| WEBSOCKETS_ENABLED | Whether or not to enable all WebSocket functionality. | Boolean | False |
| WEBSOCKETS_HEARTBEAT_ENABLED | Whether or not to enable the heartbeat ping signal. | Boolean | True |
| WEBSOCKETS_HEARTBEAT_PERIOD | The period in seconds at which to send the ping. This period doubles as the timeout used for closing an unresponsive connection. | Number | 30 |
 
| Variable           | Description                                | Type    | Default Value |
|--------------------|--------------------------------------------|---------|---------------|
| WEBSOCKETS_REST_ENABLED | Whether or not to enable the REST message handlers. | Boolean | True |
| WEBSOCKETS_REST_PATH | The URL Path at which the WebSocket will be available. | String | `/websocket` |
| WEBSOCKETS_REST_CONN_LIMIT | How many simultaneous connections are allowed. | Number | `Infinity` |
| WEBSOCKETS_REST_AUTH | What method of authentication to require for this connection. | String | `handshake` |
| WEBSOCKETS_REST_AUTH_TIMEOUT | The amount of time in seconds to wait before closing an unauthenticated connection. | Number | 30 |

| Variable           | Description                                | Type    | Default Value |
|--------------------|--------------------------------------------|---------|---------------|
| WEBSOCKETS_GRAPHQL_ENABLED | Whether or not to enable the GraphQL Subscriptions. | Boolean | True |
| WEBSOCKETS_GRAPHQL_PATH | The URL Path at which the WebSocket will be available. | String | `/graphql` |
| WEBSOCKETS_GRAPHQL_CONN_LIMIT | How many simultaneous connections are allowed. | Number | `Infinity` |
| WEBSOCKETS_GRAPHQL_AUTH | What method of authentication to require for this connection. | String | `handshake` |
| WEBSOCKETS_GRAPHQL_AUTH_TIMEOUT | The amount of time in seconds to wait before closing an unauthenticated connection. | Number | 30 |

# Security

### Authentication Modes

- Mode `PUBLIC`
    In this mode no socket authentication is required to connect to the websocket nor is it enforced for open connections.
    To authenticate after the connection has been established the `AuthHandler` can be used.
    
- Mode `HANDSHAKE`
    In this mode no authentication is required on the initial connection but the server required the very first message sent to be an authentication request for the `AuthHandler`. If no message is received within the timeout or the authentication message is not valid the connection will be closed by the server.
    
- Mode `STRICT`
    authentication is required as URL parameter on the initial connection or the connection will not be opened.
    
> NOTE: `header` and `cookie` authentication is not available in the browser WebSocket API.

### REST Authentication flow

The flow for keeping a connection authenticated is as follows:

1. (`public`/`handshake` modes) Send the initial authentication request in one of the following formats
    
    ```json
    { "type": "auth", "email": "user@email.com", "password": "string" }
    { "type": "auth", "access_token": "token" }
    { "type": "auth", "refresh_token": "token" }
    ```

2. On successful authentication you'll receive a confirmation message. This message includes a `refresh_token` when using `email/password` and `refresh_token` credentials.
   ```json
   {"type":"auth","status":"ok","refresh_token":"<A REFRESH TOKEN>"}
   ```

3. When the client receives an auth expired error (example below) a new authentication request is expected within the set `WEBSOCKETS_REST_AUTH_TIMEOUT` or the connection will be closed under `handshake` and `strict` mode.
    
    ```json
    { "type": "auth", "status": "error", "error": { "code": "TOKEN_EXPIRED", "message": "Token expired." } }
    ```

### GraphQL authentication

GraphQL puts more responsibility on the client for handling the re-authentication flow and has no supported way for us to implement it without breaking compatibility with existing clients.
Because of this only `access_tokens` are accepted for authentication at this time. When a token expires the connection will be closed with a Forbidden message signaling to the client to refresh their `access_token` and reconnect.

**Public**
```js
import { createClient } from 'graphql-ws';
const client = createClient({
	url: 'ws://localhost:8055/graphql',
	keepAlive: 30000,
});
```

**Handshake**
```js
import { createClient } from 'graphql-ws';
const client = createClient({
	url: 'ws://localhost:8055/graphql',
	keepAlive: 30000,
  connectionParams: async () => {
		return { access_token: 'MY_TOKEN' };
	},
});
```

**Strict**
```js
import { createClient } from 'graphql-ws';
const client = createClient({
	url: 'ws://localhost:8055/graphql?access_token=MY_TOKEN',
	keepAlive: 30000,
});
```

### Rate Limiter / Messenger

Currently the same rate limiter and messenger configuration set globally for Directus is used.

## Clients

There are several pre-built tools available for testing the websocket connection. But do keep in mind that a `https://` site cannot directly talk to a non-secure `ws://` socket. In order to connect some of these will require a browser extension to facilitate the cross protocol connection.
- [REST] https://www.piesocket.com/websocket-tester
- [REST] https://livepersoninc.github.io/ws-test-page/
- [GraphQL] https://altairgraphql.dev/ (using `graphql-ws`)

## Subscriptions

These subscriptions are hooked into the `items.create`,`items.update`,`items.delete` events to provide the connected clients with updates.

**REST Subscriptions**
Subscribing to collection `test` would look like this:
```js
{ 
  "type": "subscribe",
  "collection": "test",
//  "item": primaryKey,
  "query": {
    "fields": ["*"]
  },
  "uid": "unique-id",
}
```
```js
{ 
  "type": "subscribe",
  "collection": "directus_users",
//  "item": primaryKey,
  "query": {
    "fields": ["*"]
  },
  "uid": "unique-id",
}
```
> Warning: the system collections may throw many updates! be sure to select enough fields to know why it fired

Resulting in responses like:
```json
{"type":"subscription","event":"init","payload":[{"id":1,"test":"does it work?"},{"id":3,"test":"initial value"}],"uid": "unique-id"}
```
```json
{"type":"subscription","event":"update","payload":[{"id":3,"test":"updated value"}],"uid": "unique-id"}
```

**GraphQL Subscriptions**
```gql
subscription {
  test_mutated {
    id
    test
    _event <-- not an actual field
  }
}
```
```gql
subscription {
  directus_users_mutated {
    id
    first_name
    _event <-- not an actual field
  }
}
```
Resulting in:
```json
{ "data": { "test_mutated": { "id": "13", "test": "aaa", "_event": "create" } } } 
```
```json
 { "data": { "test_mutated": { "id": "13", "test": "ccc", "_event": "update" } } } 
```
```json
{ "data": { "test_mutated": { "id": "13", "test": null, "_event": "delete" } } } 
```
> note: the deleted event has the edge case that it does not actually have more data than the old ID. So requesting non-nullable information may result in errors.
> Warning: the system collections may throw many updates! be sure to select enough fields to know why it fired

## REST Items Handler
TBD

## Error Codes

The WebSocket implementation will throw a variety of errors to the client.

Error Code | Type | Reason |  
-- | -- | -- | --
INVALID_PAYLOAD | server | Unable to parse the incoming message! |  
REQUESTS_EXCEEDED | server | Rate limit reached! Try again in Xms |  
AUTH_FAILED | auth | Authentication failed. |  
AUTH_TIMEOUT | auth | Authentication timed out. |  
TOKEN_EXPIRED | auth | Token Expired. |  
INVALID_ACTION | items | Invalid action provided. |  
INVALID_PAYLOAD | items | Invalid payload in "data". |  
INVALID_QUERY | items | Invalid query provided. |  
INVALID_PAYLOAD | items | Either 'keys' or 'key' is required for a DELETE request |  
INVALID_COLLECTION | subscribe/items | The provided collection does not exists or is not accessible. |  

## Extensibility

Customizable using custom hooks, more details to come.

Fixes ENG-406

## Follow up PRs
- Auto Refresh Collection (https://github.com/directus/directus/pull/17222)

## TODO before merge
- [x] Revert triggering tests for all vendors

freekrai commented on api/src/services/graphql/index.ts:
```diff
@@ -97,6 +105,47 @@ if (env['GRAPHQL_INTROSPECTION'] === false) {
 	validationRules.push(NoSchemaIntrospectionCustomRule);
 }
 
+// DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DEMO DE
```
let's make sure we remove these before merge?

freekrai commented on app/src/websocket.ts:
```diff
@@ -0,0 +1,184 @@
+        const type = message['type'].toUpperCase()
+        const uid = 'uid' in message? Number(message['uid']) : undefined;
+    
+        console.log("messageRecieved", message)
```
also let's remove console.logs before merge?

freekrai commented on app/src/websocket.ts:
```diff
@@ -0,0 +1,184 @@
+async function authenticate(client: WebSocketClient) {
+    let token = api.defaults.headers.common['Authorization'].substring(7)
+
+    console.log("Authenticating with token", token)
```
also let's remove console.logs before merge?

freekrai commented on app/src/layouts/tabular/index.ts:
```diff
@@ -65,6 +66,35 @@ export default defineLayout<LayoutOptions, LayoutQuery>({
+						search: newSearch,
+					}
+				}, (data) => {
+					console.log("Data Changed!!!", data)
```
remove before merge?

freekrai commented on app/src/websocket.ts:
```diff
@@ -0,0 +1,184 @@
+        const response = await client.send('HANDSHAKE', {
+            access_token: token
+        })
+        console.log("Authenticated", response)
```
remove before merge?

freekrai reviewed (dismissed):
Tested locally, everything works well, just some comments about some debug console.log statements we probably want to remove before we merge this?

Nitwel commented:
Oh, this branch is very much still in development @freekrai !
Rijk probably meant by a first review to just look on the code structure in general and there are still changes to come so take every bit of code with a grain of salt at the moment. 😄

freekrai commented:
> Oh, this branch is very much still in development @freekrai ! Rijk probably meant by a first review to just look on the code structure in general and there are still changes to come so take every bit of code with a grain of salt at the moment. 😄

yeah code structure is fine, did a bunch of tests, etc and like what I saw

daawaan4U commented:
helloo, curious since i've been wanting to understand how does the implementation work but i'm struggling where to find the corresponding code for what i want to ask 😅 

how does the implementation filter and publish created/updated/deleted items to the appropriate queries subscribed? 

does it rely on the database to filter updates through polling (or reexecuting the query whenever a mutation happens) then performing a diff before sending updates to its clients similar? or does the filtering happen within the directus middleware itself? or something else?

br41nslug commented:
@daawaan4U For the subscriptions we're relying on the already existing events system (used in flows/custom hooks). As these events give partial data we do query the database to get the requested information for a certain subscription.

It's hard to point to one single place as "here is the magic" but these files are the closest to what you ask i think.
Graphql implementation: https://github.com/directus/directus/blob/6efa10428fe62c03c5d3b1dcd094e83e1598a5cd/api/src/services/graphql/subscription.ts
Rest implementation: https://github.com/directus/directus/blob/6efa10428fe62c03c5d3b1dcd094e83e1598a5cd/api/src/websocket/handlers/subscribe.ts

daawaan4U commented:
@br41nslug thanks, i'm trying to wrap my head around 😅 I'm no expert in these, but I was thinking about how the entire thing resembles hasura's implementation for their live queries (https://github.com/hasura/graphql-engine/blob/master/architecture/live-queries.md#hasura-approach). since the implementation reexecutes the query for each individual subscription to get the new data from the db, would it be possible to batch all the individual queries to the db too similar to hasura's?

br41nslug commented:
especially the GraphQL Subscriptions part yeah those will all look alike as how it's supposed to work is defined in its spec.

MasonGeloso commented:
I just want to throw it out there that this is an excellent implementation and I hope this gets merged to main soon. I work at a F500 company that is leveraging @br41nslug extension in Directus for some RT events on our frontend and it works really well with minimal overhead. 

keep up the good work! :100:

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
added emitter context

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
partial items tests

br41nslug committed:
updated items handler tests

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed test after merge

br41nslug committed:
forgot the event context

br41nslug committed:
fixed auth message parsing for graphql subscriptions

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed type strictness

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed graphql subscription bug

br41nslug committed:
bumped websocket dependencies

br41nslug committed:
touched up some dangling code

br41nslug committed:
updated itemsservice usage

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
disabled overkill logs

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
double checked environment type processing

br41nslug committed:
fixed missed capitalization

br41nslug committed:
fixed subscription payloads

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Added explicit string type casting

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Merge branch 'main' into feat/rt

azrikahar commented on api/src/websocket/controllers/base.ts:
```diff
@@ -0,0 +1,232 @@
+	protected parseMessage(data: string) {
+		let message: WebSocketMessage;
+		try {
+			message = WebSocketMessage.parse(JSON.parse(data));
```
Let's use `parseJSON` from shared package in place of `JSON.parse` 👍

azrikahar commented on api/src/websocket/controllers/rest.ts:
```diff
@@ -0,0 +1,57 @@
+	protected override parseMessage(data: string): WebSocketMessage {
+		let message: WebSocketMessage;
+		try {
+			message = JSON.parse(data);
```
commenting for parseJSON

azrikahar commented on api/src/websocket/utils/wait-for-message.ts:
```diff
@@ -0,0 +1,51 @@
+			try {
+				clearTimeout(timer);
+				client.off('message', awaitMessage);
+				resolve(JSON.parse(event.toString()));
```
commenting for parseJSON

azrikahar commented on api/src/websocket/utils/wait-for-message.ts:
```diff
@@ -0,0 +1,51 @@
+		function awaitMessage(event: RawData) {
+			let msg: WebSocketMessage;
+			try {
+				msg = WebSocketMessage.parse(JSON.parse(event.toString()));
```
commenting for parseJSON

azrikahar commented on api/src/websocket/controllers/rest.ts:
```diff
@@ -0,0 +1,57 @@
+	constructor(httpServer: httpServer) {
+		super(httpServer, 'WS REST', String(env.WEBSOCKETS_REST_PATH), {
+			mode: String(env.WEBSOCKETS_REST_AUTH).toLowerCase() as AuthMode,
+			timeout: Number(env.WEBSOCKETS_REST_AUTH_TIMEOUT) * 10000,
```
is this perhaps meant to be `1000` instead of `10000`? (similar to `WEBSOCKETS_GRAPHQL_AUTH_TIMEOUT`)

azrikahar commented on api/src/websocket/handlers/subscribe.ts:
```diff
@@ -0,0 +1,259 @@
+	 * Handle incoming (un)subscribe requests
+	 */
+	async onMessage(client: WebSocketClient, message: WebSocketSubscribeMessage) {
+		if (message.type.toUpperCase() === 'SUBSCRIBE') {
```
I'm actually not sure about the purpose/reasoning of `trimUpper`, but does this `message.type` need `trimUpper` instead of just `.toUpperCase()`? like how it is being used in:

https://github.com/directus/directus/blob/71304eb0e172245eaf2be3342e82882266129d8a/api/src/websocket/handlers/items.ts#L14

https://github.com/directus/directus/blob/71304eb0e172245eaf2be3342e82882266129d8a/api/src/websocket/controllers/base.ts#L147

br41nslug replied:
Aaah yeah the `trimUpper` util is a relic from the past 😬 It was very useful before messages were validated by `zod`.

azrikahar commented on api/src/exceptions/authentication-failed.ts:
```diff
@@ -0,0 +1,7 @@
+
+export class AuthenticationFailedException extends BaseException {
+	constructor(message = 'Authentication failed.') {
+		super(message, 401, 'AUTH_FAILED');
```
You have actually documented it the PR body! Which I assume will be carried forward to the docs later on. Marking this as resolved 😄

br41nslug committed:
removed obsolete "trimUpper" utility

br41nslug committed:
using the parseJSON utility consistently

br41nslug committed:
pinned dependencies

br41nslug committed:
parse environment variables

br41nslug committed:
fixed pnpm-lock

br41nslug committed:
GraphQL Subscriptions for all events

br41nslug committed:
fixed typo

br41nslug committed:
added event data to the graphql definition

Nitwel reviewed (commented):
There is currently a bug that if you auth with username and password, your token expires very quickly

Nitwel reviewed (commented):
Another thing we have to think through which is outside the scope of this PR, is the way we handle relational changes.

Imagine 2 collections a and b with some relation in between.
If we now change an item in collection b over an item in collection a, we will get an update on the subscription to collection a.

If instead we update the item in collection b directly, we don't get an update on the subscription. This might be very intuitive.

br41nslug committed:
fix payload for delete events

jaads commented on api/src/websocket/handlers/items.ts:
```diff
@@ -0,0 +1,98 @@
+		emitter.onAction('websocket.message', ({ client, message }) => {
+			if ((message?.type ?? '').toUpperCase() !== 'ITEMS') return;
+			try {
+				const parsedMessage = WebSocketItemsMessage.parse(message);
```
When I send `{ "type": "items", "action": "delete" }` this line throws en error which is not handled within the catch block. It's an error from Zod which basically says `"path":["collection"],"message":"Required"`.

If I add the `collection` attribute, the error is handled correct though

Nitwel commented on api/src/websocket/handlers/subscribe.ts:
```diff
@@ -0,0 +1,259 @@
+		this.messenger = getMessenger();
+		this.bindWebsocket();
+		this.bindModules([
+			// move this to constants (or maybe make it configurable)
```
Leftover food in this file.
Also a lot of console.logs later on.

br41nslug committed:
Added optional chaining for type to prevent fatal crashes on invalid messages

br41nslug committed:
fix failing on getting type from undefined

br41nslug committed:
Update api/src/websocket/exceptions.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Add proper ZodError handling

br41nslug committed:
added the zod-validation-error parser

br41nslug committed:
allow disabling the rate limiter

br41nslug committed:
Update api/src/websocket/controllers/base.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
updated starting logs

br41nslug committed:
fixed email/password expiration logic

br41nslug committed:
added tests for getMessageType

br41nslug committed:
simplified message parsing and dropped capitalization

br41nslug committed:
updated authenticate test

br41nslug committed:
switched to lower cased message.type to prevent spreading "toUpperCase" around

br41nslug committed:
cleaned up debug logs

br41nslug committed:
cast enabled config to boolean

azrikahar commented on api/src/websocket/utils/wait-for-message.ts:
```diff
@@ -0,0 +1,52 @@
+			} catch {
+				return;
+			}
+			try {
+				if (getMessageType(msg) === type) {
+					clearTimeout(timer);
+					client.off('message', awaitMessage);
+					resolve(msg);
+				}
+			} catch (err) {
+				reject(msg);
+			}
```
is this try/catch still catching a potential error (just in case it used to be, but no longer does)? or the "catch" can be "else" instead?

br41nslug committed:
Update api/src/websocket/controllers/rest.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/websocket/handlers/subscribe.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/websocket/handlers/subscribe.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/websocket/handlers/items.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/websocket/controllers/base.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/websocket/handlers/heartbeat.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug commented on api/src/websocket/controllers/graphql.ts:
```diff
@@ -0,0 +1,89 @@
+
+				const state = await authenticateConnection(BasicAuthMessage.parse(msg.payload));
+				this.server.emit('connection', ws, state);
+				ws.send(JSON.stringify({ type: 'connection_ack' }));
```
I that would work however the intention of that `fmtMessage` function is to add extra data in an organized way. I did not use the function here because GraphQL does not play well with extra data in this response.

azrikahar replied:
Thanks for the clarification! I was indeed curious whether it's intentional in GraphQL controller since it's the only one, so I was hesitant on the need for it. Makes sense, marking this as resolved 👍

br41nslug committed:
Suggested fixes by Azri

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
removed redundant try-catch

br41nslug committed:
fixed authentication timeout
added returning the refresh token when authenticating

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
updated pnpm lock after merge

br41nslug committed:
Fixed authentication modes for GraphQL according to best practices

azrikahar committed:
Merge branch 'main' into feat/rt

azrikahar committed:
implement useFakeTimers in heartbeat unit test

azrikahar committed:
implement useFakeTimers in items unit test

azrikahar commented:
I've updated the handlers unit test (https://github.com/directus/directus/pull/14737/commits/cd6e2e31153a3a0297890cb9ddd73113b7996185 and https://github.com/directus/directus/pull/14737/commits/884c2151861fef441bc611c6a7caa187286c72eb) and replaced the delay function with vitest `useFakeTimers` instead, hopefully that's more stable 👍

Seems to improve `pnpm -F directus test websocket/handlers` time taken from roughly 6 seconds to 2 seconds locally, and it should be less hardware dependent (as the previous +10ms delay will still sometimes fail for me after testing it multiple times) 🙌

br41nslug committed:
Update api/src/services/server.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug commented on api/src/websocket/controllers/base.ts:
```diff
@@ -0,0 +1,242 @@
+		authentication: {
+			mode: AuthMode;
+			timeout: number;
+			verbose: boolean;
```
This was toggling the authentication response which was made obsolete by recent changes 👍

br41nslug commented on api/src/websocket/handlers/subscribe.ts:
```diff
@@ -0,0 +1,253 @@
+			const { client } = subscription;
+			try {
+				client.accountability = await refreshAccountability(client.accountability);
+				const schema = await getSchema();
```
Good suggestion, this made more sense when the `getSchema` function was still using the `accountability` to fetch the appropriate schema.

azrikahar commented on api/src/websocket/handlers/subscribe.ts:
```diff
@@ -0,0 +1,253 @@
+				// remove the subscription if it already exists
+				this.unsubscribe(client, subscription.uid);
+
+				let data: Record<string, any>;
+				if ('item' in subscription) {
+					data = await this.getSinglePayload(subscription, accountability, schema);
+				} else {
+					data = await this.getMultiPayload(subscription, accountability, schema);
+				}
```
I believe this if-else is similar in nature with the ternary here (aside from the optional 4th param `event`):

https://github.com/directus/directus/blob/a9fe1ad6c818c0a5120d8a90b6f7efbbc31923b5/api/src/websocket/handlers/subscribe.ts#L139-L142

would it be possible to make both consistently if-else, or consistently ternary logic? Personally am leaning towards ternary since `data` isn't mutated here, if I'm not mistaken, but no qualms with if-else 👍

br41nslug commented on api/src/services/graphql/subscription.ts:
```diff
@@ -0,0 +1,82 @@
+		const { fields } = self.getQuery({}, selections, {});
+		for await (const payload of messages.subscribe(event)) {
+			const eventData = payload as Record<string, any>;
+			const schema = await getSchema();
```
In this case it cannot be moved outside the loop or the schema will never be re-fetched while Directus is running.

azrikahar replied:
Ohhh gotcha, thanks for clarifying! Marking this as resolved 👍

br41nslug commented on api/src/services/websocket.ts:
You're not wrong that it's not actively used at this time. However this is the `WebsocketService` intended for extensions to use.

azrikahar replied:
Makes perfect sense! 🤯

br41nslug committed:
removed obsolete authentication.verbose toggle

br41nslug committed:
added email flag to message validation

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
switched to ternary for consistency

br41nslug committed:
moved getSchema out of for loop

br41nslug committed:
added singleton logic to items handler

Nitwel commented:
Tested authentication with username/pw and after refreshing the connection once, it does not seem to close the connection after refreshing with an invalid refresh token.
```
[20.03.2023 11:36.27.666] [LOG]   Connecting to server...
[20.03.2023 11:36.27.683] [LOG]   WebSocket Client Connected
[20.03.2023 11:36.27.684] [LOG]   Sending: {"type":"auth","email":"admin@example.com","password":"pw"}
[20.03.2023 11:36.28.204] [LOG]   Received: '{"type":"auth","status":"ok","refresh_token":"rmHJLhUK0Vw8vhnSFJKUalOzgh9eBCfDKyJAbQyymHx3bsKnRLMqFze_UjrRwPhY"}'
[20.03.2023 11:36.28.205] [LOG]   Logged in successfully
[20.03.2023 11:37.27.011] [LOG]   Received: '{"type":"auth","status":"error","error":{"code":"TOKEN_EXPIRED","message":"Token expired."}}'
[20.03.2023 11:37.27.011] [LOG]   Sending: {"type":"auth","refresh_token":"rmHJLhUK0Vw8vhnSFJKUalOzgh9eBCfDKyJAbQyymHx3bsKnRLMqFze_UjrRwPhY"}
[20.03.2023 11:37.27.022] [LOG]   Received: '{"type":"auth","status":"ok","refresh_token":"k9MJ0KYUX0awhSjqsd1SwPTcnJtnDLlptW9lCV4R-mxK6aNi1PbvfzVPr-yJekrJ"}'
[20.03.2023 11:38.27.012] [LOG]   Received: '{"type":"auth","status":"error","error":{"code":"TOKEN_EXPIRED","message":"Token expired."}}'
[20.03.2023 11:38.27.012] [LOG]   Sending: {"type":"auth","refresh_token":"rmHJLhUK0Vw8vhnSFJKUalOzgh9eBCfDKyJAbQyymHx3bsKnRLMqFze_UjrRwPhY"}   
[20.03.2023 11:38.27.020] [LOG]   Received: '{"type":"auth","status":"error","error":{"code":"AUTH_FAILED","message":"Authentication failed."}}'
```

br41nslug committed:
close the socket after failed auth for non-public connections

br41nslug commented:
Because @azrikahar pointed out an issue with subscribing to `directus_fields` using REST i've disabled subscribing to system collections for this PR in favor of solving this for both GraphQL subscriptions and REST in a follow up PR.

br41nslug committed:
disabled system collections for rest subscriptions

rijkvanzanten commented:
> i've disabled subscribing to system collections for this PR in favor of solving this for both GraphQL subscriptions and REST in a follow up PR.

@br41nslug Nahhhhhhhhh lets add that in here. Being able to subscribe to users/files among others will be very important to have 😬

br41nslug commented:
> Being able to subscribe to users/files among others will be very important to have 😬

Absolutely agree, the hope was to make it less of a reviewing nightmare by reviewing that extra implementation separately

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
re-ran pnpm i

br41nslug committed:
allow for multiple subscripitions in the memory messenger

br41nslug committed:
- fixed system collection subscriptions
- abstracted hook message bus
- fixed graphql horizontal scaling

br41nslug committed:
remove logic from root context for tests

licitdev committed:
Merge branch 'main' into feat/rt

Nitwel committed:
fix reading created item

Nitwel commented:
Made some small changes, let me know if you're fine with them!

Nitwel reviewed (approved):
This is looking good to me know! I was able to test and confirm the following:
- CRUD implementation
- REST Subscriptions
- Authentication modes: public, handshake, strict
- Authentication refresh
- Heartbeat signal
- Permissions
- Message rate limiting and connection limiting
- Horizontal scaling with the Messenger

I did not test the following:
- GraphQL Subscriptions (as I was not able to get that to work on my side)
- Code Quality (as Azri already did an amazing job covering that)

br41nslug committed:
fix linter

azrikahar commented on api/src/websocket/utils/get-expires-at-for-token.test.ts:
```diff
@@ -0,0 +1,21 @@
+		const result = getExpiresAtForToken(token);
+		expect(result).toBe(null);
+	});
+	test('Returns experes field for jwt with exp as number', () => {
```
is `experes` meant to be `expires`/`expiresAt` here?

azrikahar commented on api/src/websocket/utils/get-expires-at-for-token.ts:
```diff
@@ -0,0 +1,9 @@
+import jwt from 'jsonwebtoken';
+
+export function getExpiresAtForToken(token: string): number | null {
+	const decoded = jwt.decode(token);
```
I believe similarly in `getAccountabilityForToken`, should we only get the decoded `exp` if it's a Directus issued token? 🤔

Although this is very much an edge case and requires a user to be able edit their own static token, but it seems to be technically possible to use a generic JWT such as `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2ODE1NTQ4MDB9.91LkM-af_GOTpawYNPRbVK9dfqMJyA_J7UghPLEz8W8` (the payload is `{ "exp": 1681554800 }`) to "artificially" extend the auth timer duration.

br41nslug replied:
For this utility specifically it shouldn't matter where the JWT token comes from but the token should definitely be checked for validity before being handed to to this function.

The check for whether it is a valid Directus token happens in the `getAccountabilityForToken` function that will throw an error if not validated https://github.com/directus/directus/blob/12085bacbabacc57a046534236fd65189533bc80/api/src/websocket/controllers/base.ts#L88-L89

azrikahar replied:
> Wouldn't this technically be lowering the expire duration of a static token because they dont expire at all

Ahh this is definitely more accurate than what I was (wrongly) thinking! 

The concern about valid Directus token was only about trying not to "accidentally" extract `exp`. In that case, will that count as a potential issue? Though understandably this might be very unlikely for a user to just so happen to set a custom JWT _with exp_ as their static token. Feel free to resolve this if it's too much of an edge case, especially in terms of MVP 👍

azrikahar commented on api/src/websocket/controllers/base.ts:
```diff
@@ -0,0 +1,246 @@
+			client.auth_timer = null;
+		}
+		if (!client.expires_at) return;
+		const expiresIn = Math.max(client.expires_at * 1000 - Date.now(), 10);
+		client.auth_timer = setTimeout(() => {
+			client.accountability = null;
+			client.expires_at = null;
+			handleWebsocketException(client, new TokenExpiredException(), 'auth');
+			waitForMessageType(client, 'auth', this.authentication.timeout).catch((msg: WebSocketMessage) => {
+				const error = new WebSocketException('auth', 'AUTH_TIMEOUT', 'Authentication timed out.', msg?.uid);
+				handleWebsocketException(client, error, 'auth');
+				if (this.authentication.mode !== 'public') {
+					client.close();
+				}
+			});
+		}, expiresIn);
```
I'm not sure what could be the safe guard yet, but it seems like when `expiresIn` is calculated to be higher than 32 bit integer limit (`2,147,483,647`), the setTimeout here will throw overflow warning and default to `1` instead:

![](https://user-images.githubusercontent.com/42867097/226635733-205b4a67-21a7-40d0-ac77-7f17a3259513.png)

which then cause said auth attempt to immediately receive TOKEN_EXPIRED error.

To reproduce this, try setting `ACCESS_TOKEN_TTL` to `25d` (because this converts to slightly higher than 32 bit int limit in milliseconds) or higher and authenticate with email/password.

br41nslug replied:
That is a very good catch, it would have never occurred to me that someone might set their `ACCESS_TOKEN_TTL` to something ridiculous like 20+days 😂 may as well just work with static access tokens at that point.

But bad security configurations aside this definitely should not overflow the integer 🤔

br41nslug committed:
typo and extra safe guard suggested by azri

br41nslug committed:
prevent setting long timeouts in favor of a shared interval

azrikahar commented:
### Potential issue

For REST subscriptions, it seems like subscribing a second collection is causing the client to no longer receive events for the first collection. 

For GraphQL subscriptions, subscribing two collections seems to work as expected.

#### Example test case with REST subscriptions

1. Create collection `a` with a string field `name`.
2. Create collection `b` with a string field `name`.
3. Use REST subscription on collection `a` with a ws client.
4. Create first item in collection `a`, the ws client receives `{"type":"subscription","event":"create","payload":[{"id":1,"name":"A1"}]}`.
5. using the same ws client, additionally subscribe to collection `b`.
6. Create second item in collection `a`, the ws client doesn't receive any message for it anymore.
7. Create first item in collection `b`, the ws client receives `{"type":"subscription","event":"create","payload":[{"id":2,"name":"B1"}]}`.

---

### Additional Thoughts

Currently for REST subscriptions, we might not be able to distinguish what collection it is, especially when we're subscribed to multiple collections (unless REST isn't meant to be able to). I'm not sure do we, or can we, make that info available. Sort of how action hooks has "meta" such as the collection.

br41nslug commented:
> Currently for REST subscriptions, we might not be able to distinguish what collection it is, especially when we're subscribed to multiple collections (unless REST isn't meant to be able to). I'm not sure do we, or can we, make that info available. Sort of how action hooks has "meta" such as the collection, but understandably that may require changing the current spec of the responses.

This is what the `uid` is for so you can distinguish between the responses of multiple subscriptions.

br41nslug committed:
prevent unsubscribing all existing subscriptions when omitting "uid"

br41nslug commented:
@azrikahar I was wondering why i could not reproduce this issue 🤦 Turns out i was always setting the `uid` in my tester.
![image](https://user-images.githubusercontent.com/9389634/226877225-e5ebcb8b-45bb-4b51-bc17-ee38eb921243.png)

While adding a `uid` effectively solved it there was also an unintended effect of clearing all your subscriptions when the `uid` was omitted.

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
- extracted getService utility
- block system collections mutation in the items handler
- implemented the correct services for system collections

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
allow numeric uid's to be used

br41nslug committed:
fixed the types for numeric uid's to be used

br41nslug committed:
added missing await's

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed type imports after merge

rijkvanzanten committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
removed unused imports

azrikahar commented on api/src/websocket/controllers/hooks.ts:
```diff
@@ -0,0 +1,130 @@
+}
+
+function registerRelationsHooks() {
+	// exception for relation updates that don't report `directus_relations` as being the collection
```
```suggestion
	// exception for relation hooks that don't report `directus_relations` as being the collection
```

(just to disambiguate it with `update` event specifically)

azrikahar commented on api/src/services/graphql/index.ts:
```diff
@@ -1055,6 +1046,19 @@ export class GraphQLService {
 					});
 				}
+
+				const eventName = `${collection.collection}_mutated`.toUpperCase();
+				const subscriptionName = camelCase(eventName);
+				const subscriptionType = ReadCollectionTypes[collection.collection].clone(collection.collection + '_mutation');
+				subscriptionType.addFields({
+					event: GraphQLString,
+				});
+				schemaComposer.Subscription.addFields({
+					[subscriptionName]: {
+						type: subscriptionType,
+						subscribe: createSubscriptionGenerator(self, eventName, subscriptionName),
+					},
+				} as any);
```
Seems like because the for loop wrapping this subscription logic has this line at the top:

https://github.com/directus/directus/blob/d23f2c0c1a538c30203d26e14208c962d02c9bc5/api/src/services/graphql/index.ts#L803

`directus_collections`, `directus_fields`, and `directus_relations` are currently not available via GraphQL Subscriptions:

https://github.com/directus/directus/blob/d23f2c0c1a538c30203d26e14208c962d02c9bc5/api/src/services/graphql/index.ts#L94-L97

br41nslug replied:
You cannot query these with GraphQL either. This is a limitation of the current GraphQL implementation and i consider this out of scope for this PR.

azrikahar reviewed (commented):
- [x] ~~The latest fix for system collections works like a charm for REST subscriptions! A minor GraphQL subscriptions bug is a couple of those system collections are not avaiable, as mentioned in [this review comment](https://github.com/directus/directus/pull/14737#discussion_r1145721058).~~ This is invalid as they are not query-able as Tim mentioned.

- [x] ~~When WEBSOCKETS_GRAPHQL_AUTH is handshake, attempting to connect to GraphQL subscriptions without setting any access token is correctly getting rejected. However the error message is the shared/generic error:~~ 

    ```json
    {
      "type": "error",
      "status": "error",
      "error": {
        "code": "INVALID_PAYLOAD",
        "message": "Validation error: Required at \"email\"; Required at \"password\", or Required at \"access_token\", or Required at \"refresh_token\""
      }
    }
    ```
    
    ~~The mentions of email/password/refresh_token might potentially confuse users when they can only use `access_token` for GraphQL subscriptions. 🤔~~

- [ ] ~~Should we check `WEBSOCKETS_REST_AUTH` and `WEBSOCKETS_GRAPHQL_AUTH` for valid values (public, handshake, strict)? Technically an invalid value seems to works like "public" (if I'm not mistaken?), but it might be overall safer to validate that they will only be one of valid values.~~

br41nslug committed:
Update api/src/websocket/controllers/hooks.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/websocket/controllers/hooks.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
Update api/src/messenger.ts

Co-authored-by: Azri Kahar <42867097+azrikahar@users.noreply.github.com>

br41nslug committed:
improved error for graphql subscriptions

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed TS Modernization conflicts

br41nslug committed:
fixed TS Modernization conflicts

itzomen commented:
> I just want to throw it out there that this is an excellent implementation and I hope this gets merged to main soon. I work at a F500 company that is leveraging @br41nslug extension in Directus for some RT events on our frontend and it works really well with minimal overhead.
> 
> keep up the good work! 💯

Same, this implementation is 🔥 fire
Thanks a bunch @br41nslug

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed conflicts after merge

jaads commented on api/src/websocket/controllers/rest.ts:
```diff
@@ -0,0 +1,55 @@
+	constructor(httpServer: httpServer) {
+		super(httpServer, 'WS REST', String(env['WEBSOCKETS_REST_PATH']), {
+			mode: String(env['WEBSOCKETS_REST_AUTH']).toLowerCase() as AuthMode,
+			timeout: Number(env['WEBSOCKETS_REST_AUTH_TIMEOUT']) * 1000,
```
shall we move the logic of the above two lines into the abstract class and then just pass the names of the environment variables to it? this would remove some duplicated code like the casting, the lowercase function call and the multiplication within is currently done in both concrete classes 🤔

jaads replied:
actually also for the logger statement at the end of the constructor. if you would move this statement out of the concrete class and use the `name` variable in the abstract class, you could even get rid of the dependency to `../../env` in the concrete classes

br41nslug replied:
Hmmm interesting suggestion 🤔 We could probably, abstract this into a "getConfig" function but im not 💯% sure that it's worth the abstraction for these 2 lines.
This was originally setup with the idea that these configs could diverge at some point but that never happened, will definitely have a quick look if a sensible abstraction can be done without too much hassle 😄

jaads replied:
compared to what you all did in this PR this shouldn't be too much hassle 😆😆  very nice work by the way!!

br41nslug committed:
removed unused name property

br41nslug committed:
abstracxted environment configuration

br41nslug committed:
respond to ping messages when heartbeat disabled

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
something something merge

br41nslug committed:
moved toBoolean to it's own util file

licitdev commented on api/src/services/websocket.ts:
I believe this file can be deleted as it is unused.

licitdev commented on api/src/exceptions/index.ts:
```diff
@@ -19,3 +19,4 @@ export * from './unprocessable-entity';
 export * from './unsupported-media-type';
 export * from './user-suspended';
 export * from './unexpected-response';
+export * from './authentication-failed';
```
Should remove this export.

```suggestion
```

licitdev commented on api/src/exceptions/authentication-failed.ts:
I believe this file can be deleted as it is unused.

licitdev reviewed (commented):
Should we use the uppercase `S`, from `Websocket` to `WebSocket` for consistency? Noticed there were more usages for `WebSocket`.

LasseRosenow commented:
> Should we use the uppercase S, from Websocket to WebSocket for consistency? Noticed there were more usages for WebSocket.

I think that would be good, as this is the official writing see:
https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API?retiredLocale=de

br41nslug committed:
replaced old socket naming

br41nslug committed:
removed old exception

licitdev commented on api/src/websocket/controllers/base.ts:
```diff
@@ -0,0 +1,294 @@
+		this.checkClientTokens();
+		registerWebsocketEvents();
+	}
+	protected getEnironmentConfig(configPrefix: string): {
```
```suggestion
		const { endpoint, authentication, maxConnections } = this.getEnvironmentConfig(configPrefix);
		this.endpoint = endpoint;
		this.authentication = authentication;
		this.maxConnections = maxConnections;
		this.rateLimiter = this.getRateLimiter();

		httpServer.on('upgrade', this.handleUpgrade.bind(this));
		this.checkClientTokens();
		registerWebsocketEvents();
	}
	protected getEnvironmentConfig(configPrefix: string): {
```

br41nslug committed:
fixed typo

licitdev commented on api/src/env.ts:
```diff
@@ -282,6 +285,18 @@ const defaults: Record<string, any> = {
+	WEBSOCKETS_GRAPHQL_AUTH_TIMEOUT: 10,
+	WEBSOCKETS_GRAPHQL_PATH: '/graphql',
+	WEBSOCKETS_HEARTBEAT_ENABLED: true,
+	WEBSOCKETS_HEARTBEAT_FREQUENCY: 30,
```
```suggestion
	WEBSOCKETS_HEARTBEAT_PERIOD: 30,
```

licitdev commented on api/src/services/server.ts:
```diff
@@ -94,6 +94,32 @@ export class ServerService {
+					: false;
+
+				info['websocket'].heartbeat = env['WEBSOCKETS_HEARTBEAT_ENABLED']
+					? env['WEBSOCKETS_HEARTBEAT_FREQUENCY']
```
```suggestion
					? env['WEBSOCKETS_HEARTBEAT_PERIOD']
```

licitdev commented on api/src/websocket/handlers/heartbeat.test.ts:
```diff
@@ -0,0 +1,87 @@
+	const actual = (await vi.importActual('../../env')) as { default: Record<string, any> };
+	const MOCK_ENV = {
+		...actual.default,
+		WEBSOCKETS_HEARTBEAT_FREQUENCY: 1,
```
```suggestion
		WEBSOCKETS_HEARTBEAT_PERIOD: 1,
```

licitdev commented on api/src/websocket/handlers/heartbeat.ts:
```diff
@@ -0,0 +1,75 @@
+import env from '../../env';
+import { toBoolean } from '../../utils/to-boolean';
+
+const HEARTBEAT_FREQUENCY = Number(env['WEBSOCKETS_HEARTBEAT_FREQUENCY']) * 1000;
```
```suggestion
const HEARTBEAT_FREQUENCY = Number(env['WEBSOCKETS_HEARTBEAT_PERIOD']) * 1000;
```

licitdev reviewed (commented):
A slight name change from frequency to period for heartbeat to better suit the usage.

br41nslug committed:
Update api/src/env.ts

Co-authored-by: ian <licitdev@gmail.com>

br41nslug committed:
Update api/src/websocket/handlers/heartbeat.test.ts

Co-authored-by: ian <licitdev@gmail.com>

br41nslug committed:
Update api/src/websocket/handlers/heartbeat.ts

Co-authored-by: ian <licitdev@gmail.com>

br41nslug committed:
Update api/src/services/server.ts

Co-authored-by: ian <licitdev@gmail.com>

br41nslug committed:
fixed for linter

azrikahar commented on api/src/services/server.ts:
```diff
@@ -94,6 +94,30 @@ export class ServerService {
 		}
 
+		if (this.accountability?.user) {
+			if (env['WEBSOCKETS_ENABLED']) {
```
thoughts on moving the whole `if (env['WEBSOCKETS_ENABLED'])` into the same `if (this.accountability?.user)` in line 54 above?

br41nslug replied:
the outer `if` is there to guard against the websocket key being set when requesting server info unauthenticated. if we were to merge those it would always be set.

azrikahar replied:
I was wondering about moving it into the existing authenticated block here: https://github.com/directus/directus/blob/80fb6010b032888c2b3183540eff4b0253098fd7/api/src/services/server.ts#L54-L75, so we'll have the same "authenticated-only keys" in the same IF block 👍 sorry for the confusion 😄

br41nslug replied:
We definitely could do that, i dont have strong oppinions either way for this change 😬

azrikahar replied:
Likewise! But let's table this for now since it isn't really something crucial 😄

azrikahar committed:
add server_info_websocket in graphql

br41nslug committed:
Merge branch 'main' into feat/rt

jaads reviewed (approved):
- The implementation looks very well integrated, nice work!
- The usage of classes for the controller seem to be appropriate. I like the restricted visibility of the class members
- Unit tests are also looking good. Too bad that the websocket controllers cannot be unit tested easily but the integration test should cover those parts. I tested the CRUD functionality manually and it worked as expected!
- A few more comments would be nice though or some other kind of developer documentation. But that's is open to discussion.

Overall veery nice work 👍 🎉

br41nslug committed:
Merge branch 'main' into feat/rt

licitdev committed:
Add base REST websocket tests

u12206050 commented:
Don't know if I am missing additional config, but I have run into a problem when using this on Cloud run across multiple containers. It seems some events aren't triggered and I think it is either because the event happened in another container, or that the container that the websocket was connected to has been restarted. @br41nslug any ideas?

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
do merge things

br41nslug committed:
fixing things

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed failing unit test

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

licitdev committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Merge branch 'main' into feat/rt

licitdev committed:
Merge branch 'main' into feat/rt

licitdev committed:
Update dependencies

licitdev committed:
Move tests

licitdev committed:
Update lockfile

licitdev committed:
Use new paths when spawning

br41nslug committed:
return websockets to opt-in

licitdev committed:
Enable websockets for tests

licitdev committed:
Test with ephemeral access token

br41nslug committed:
no camelcasing gql subscriptions

br41nslug committed:
use underscore for gql event

licitdev committed:
Remove unused import

rijkvanzanten committed:
Merge branch 'main' into feat/rt

licitdev committed:
Add base GraphQL subscription tests

licitdev committed:
Fix accidental comment

licitdev committed:
Add some relational tests

licitdev committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Organize imports

Using VS Code's default organize import

rijkvanzanten committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Run ESlint formatting

rijkvanzanten commented on api/src/services/websocket.ts:
```diff
@@ -0,0 +1,32 @@
+	on(event: 'connect' | 'message' | 'error' | 'close', callback: ActionHandler) {
+		emitter.onAction('websocket.' + event, callback);
+	}
+	off(event: 'connect' | 'message' | 'error' | 'close', callback: ActionHandler) {
```
hihi #18197

rijkvanzanten committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
One more opinionated formatting change

rijkvanzanten committed:
Merge branch 'main' into feat/rt

rijkvanzanten committed:
Formatting

licitdev committed:
Fix message sequence not in order

licitdev committed:
Remove relational batch update tests

licitdev committed:
Test horizontal scaling

br41nslug committed:
using toboolean util for server_info

br41nslug committed:
removed unneeded type cast

br41nslug committed:
found the gql request type

br41nslug committed:
extra usage of the toBoolean util

br41nslug commented on api/src/services/server.ts:
```diff
@@ -94,6 +94,30 @@ export class ServerService {
+					  }
+					: false;
+
+				info['websocket'].graphql = env['WEBSOCKETS_GRAPHQL_ENABLED']
```
Sure why not, as you can see i just kept stacking identifiers after the `WEBSOCKETS_` prefix 😂

br41nslug committed:
merge the authentication middleware and get-account-for-token util

br41nslug committed:
updated utility test

br41nslug committed:
fixed middleware unit test

licitdev committed:
Add return

licitdev committed:
Remove user filtering and close conns

licitdev committed:
Fix reused accountability

br41nslug committed:
fixed failing util test

br41nslug committed:
added subscription unit tests

br41nslug committed:
added missing mock

br41nslug committed:
trigger workflow

br41nslug committed:
Revert "trigger workflow"

This reverts commit 4f544b0c1b0f1df6420ea38d2e8036a47db10b67.

licitdev committed:
Trigger testing for all vendors

br41nslug committed:
add unsubscription confirmation

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

licitdev committed:
Wait for unsubscription confirmation

licitdev committed:
Fix incorrect sending of unsubscription confirmation

br41nslug committed:
updated ubsubscribe logic

licitdev committed:
Update count for unsubscription message

licitdev committed:
Fix sequence for UUID pktype in MSSQL

phazonoverload committed:
Merge branch 'main' into feat/rt

licitdev committed:
Increase auth timeout

licitdev committed:
Add start index when getting messages

licitdev committed:
Fix subscription retrieval and cast uid to string

licitdev committed:
Remove nested ternary

licitdev committed:
Revert "Increase auth timeout"

This reverts commit 10707409c4816d0698f033fe51cd59810dd9649c.

phazonoverload committed:
Merge branch 'main' into feat/rt

licitdev committed:
Terminate connection instead of close

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed merge

br41nslug committed:
re-added missing packages after merge resolve

br41nslug committed:
fixed type imports

br41nslug commented:
> Settings -> Role -> Custom Permissions -> Rule -> $CURRENT_USER doesn't work, and an empty array is returned. If replaced with a real user.ID works. ![image](https://user-images.githubusercontent.com/3495993/234347961-70ed6c5d-b059-4f6e-a22f-44f3d7810d6d.png)

How does this have anything to do with this PR?

br41nslug commented:
This shouldnt work on any branch if the m2m returns more than 1 item unless i missed a fix for this on `main`.

br41nslug commented:
> https://github.com/directus/directus/blob/19b7c77c035416343b0867b75193e89689087071/api/src/utils/get-permissions.ts#L151
> 
> I believe this is responsible to access $CURRENT_USER, and I couldn't find any references from within the websockets.
> Variables seem to be working on rest but not websocket rest.

Interesting 🤔 i'll double check tomorrow

paescuj reviewed (dismissed):
@estheragbaje and I just noticed that the `STRICT` mode isn't working because the `access_token` is expected to be in the message but comes as a parameter in the URL: https://github.com/directus/directus/pull/14737/files#diff-42cb5496adcc3c5ad7bb34721c31d05301371871949d635d1e1aed045b41ad99R58

Also it took us some time to identify the issue because the message is swallowed on the client side, probably because we aren't sending the error message in the correct format 🤔

The error we received in the client:
```
Error: "error" message expects the 'id' property to be a string, but got undefined
    at validateMessage (node_modules/graphql-ws/lib/common.mjs:122:23)
```

Sounds like we need to add an `id` property to the message in `handleWebSocketException`?

---

I think this will be fixed anyway but just to have mentioned it: Even if the error would have been sent correctly to the client, the message from `zod` wouldn't have been very helpful because `message['payload']` is undefined:
```
{"type":"error","status":"error","error":{"code":"INVALID_PAYLOAD","message":"Validation error: Required"}}
```

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Create lazy-cows-happen.md

Added changeset

paescuj committed:
Minor bump for "directus" package as well

br41nslug commented:
@cliqer I've tested the `$NOW` and `$CURRENT_USER` dynamic vars and they seem to be working as expected in both the `items` and `subscription` handler. Resolving of these variables is handled by the ItemsServices used.

br41nslug committed:
fixed "strict" auth mode for graphql subscriptions

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
removed nested ternary

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

licitdev committed:
Add websocket tests to sequential flow

licitdev committed:
Disable pressure limiter for blackbox tests

licitdev committed:
Merge branch 'main' into feat/rt

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fix merge

phazonoverload committed:
WebSockets Documentation (#18254)

* Small repsonsive styling fix on Card

* REST getting started guide

* Authentication guide

* REST subscription guides

* JS Chat guide

* Sidebar websocket guides section

* Added config options

* Respoinding to brainslug's review

* Fixed incorrect header on guides/rt/subs

* Fixed spellchecker

* Correct full code example on guides/rt/chat/js

* Fixed JS chat tut

* Order of steps in js chat guide updated for easier following-along

* Realtime chat Vue Guide

* feat: create react.js file

* feat: add set up for directus project

* docs: create react boilder plate

* docs: initialize connection

* docs: set up submission methods

* docs: establish websocket connection

* docs: subscribe to messages

* docs: create new messages

* docs: display new messages

* docs: display historical messages

* docs: next steps

* docs: full code sample

* docs: clean up

* docs: add name to contributors

* docs: add react card

* docs: updates to react chat

* Added live poll result guide

* docs: intro

* docs: before you begin

* docs: install packages

* docs: authenticate connection

* docs: query and mutation

* docs: utilize hooks

* docs: subscribe to changes

* docs: create helper functions

* docs: display messages

* docs: summary

* docs: full sample code

* chore: add card for webscockets with graphql

* docs: intro

* docs: subscribe to changes

* docs: handling changes

* docs: crud operations

* docs: unsubscribing from changes

* docs: updates

* chore: add card

* chore: updates to graphql docs

* chore: updates to getting started

* chore: updates to subscription

* chore: updates to real chat guide

* Added WebSockets Operations Guide

* Consistent titles

* Contributors component for docs

* Triggering Netlify

* Add operations to sidebar

* Fix operations link

* Small formatting changes

* Clarity around property values

* Removed unused values in Contributors component

* Prompt for default choice

* Tabs & lowercase doctypes

* Semicolons

* Event overwerites -> event listeners

* Spacing

* Flipped order of websockets guide to match GQL

---------

Co-authored-by: Esther Agbaje <folasadeagbaje@gmail.com>
Co-authored-by: Rijk van Zanten <rijkvanzanten@me.com>

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
fixed typo

br41nslug committed:
removed unused import

maartenvn commented:
I think it would be a great addition to allow subscriptions with GraphQL over http using [Server-Send Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events).

The big benefit of this is that you are able to use a single GraphQL client to handle queries, mutations and subscriptions.

This is the default behaviour in [GraphQL Yoga Server](https://the-guild.dev/graphql/yoga-server/docs/features/subscriptions). There is an option in the configuration to allow for changing which way subscriptions are handled (WebSockets or Server-Send Events).

Perhaps we should switch to GraphQL Yoga over the vanilla GraphQL package, as it offers a great set of features and offers great performance. This would also unlock the GraphQLI interface for testing GraphQL queries in development, without the need for a 3th party service.

But that is would be better discussed in a separate discussion, if this is something that would be considered.

br41nslug commented:
> I think it would be a great addition to allow subscriptions with GraphQL over http using [Server-Send Events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events).
> 
> The big benefit of this is that you are able to use a single GraphQL client to handle queries, mutations and subscriptions.
> 
> This is the default behaviour in [GraphQL Yoga Server](https://the-guild.dev/graphql/yoga-server/docs/features/subscriptions). There is an option in the configuration to allow for changing which way subscriptions are handled (WebSockets or Server-Send Events).
> 
> Perhaps we should switch to GraphQL Yoga over the vanilla GraphQL package, as it offers a great set of features and offers great performance. This would also unlock the GraphQLI interface for testing GraphQL queries in development, without the need for a 3th party service.
> 
> But that is would be better discussed in a separate discussion, if that is something that would be considered.

That sounds like a great [feature request](https://github.com/directus/directus/discussions/categories/feature-requests) ;) 🙌 🚀 However this PR is specifically aimed at implementing WebSockets.

br41nslug committed:
added tests for "to-boolean" and "exceptions"

br41nslug committed:
added websocket service tests

br41nslug committed:
quote environment variable to satisfy dictionary

br41nslug committed:
Merge branch 'main' into feat/rt

br41nslug committed:
GraphQL Subscriptions update (#18804)

* updated graphql subscription structure

* updated graphql examples

* Create hungry-points-rescue.md

* using `key` instead of `ID` on the toplevel

* removed changeset

* fixed the graphql type after the rename to `key`

* retrun data null for delete events to prevent non-nullable gql error

* updated missed ID reference in the docs

* updated missed ID reference in the docs

* renamed "payload" to "data" in the REST Subscription response

* fixed missed reference to payload

* added optional event filter for REST subscriptions

* updated docs for event filter

* Update docs/guides/real-time/subscriptions/websockets.md

Co-authored-by: ian <licitdev@gmail.com>

---------

Co-authored-by: ian <licitdev@gmail.com>

licitdev committed:
Merge branch 'main' into feat/rt

br41nslug committed:
added messenger unit test

br41nslug committed:
Merge branch 'feat/rt' of https://github.com/directus/directus into feat/rt

br41nslug committed:
always send subscription confirmation

licitdev committed:
Add event to subscription options

licitdev committed:
Update tests

licitdev committed:
Merge branch 'main' into feat/rt

paescuj committed:
Merge branch 'main' into feat/rt

licitdev committed:
Add tests for event filtering

licitdev committed:
Revert testing for all vendors

paescuj committed:
Remove obsolete console comment

paescuj commented on api/src/websocket/controllers/graphql.ts:
```diff
@@ -0,0 +1,121 @@
+			schema: async (ctx) => {
+				const accountability = ctx.extra.client.accountability;
+
+				// for now only the items will be watched in the MVP system events tbd
```
```suggestion
				// for now only the items will be watched, system events tbd
```

paescuj committed:
Update comment

paescuj commented on docs/guides/real-time/chat/javascript.md:
````diff
@@ -0,0 +1,298 @@
+Finally, create event listeners which are triggered on the form submissions:
+
+```js
+document.querySelector('#login').addEventListener('open', function(event) {
````
```suggestion
document.querySelector('#login').addEventListener('submit', function(event) {
```

paescuj committed:
Correct event in JS WS guide

paescuj commented on docs/guides/real-time/chat/javascript.md:
```diff
@@ -0,0 +1,298 @@
+### Set Up Your Directus Project
+You will need a Directus project. If you don’t already have one, the easiest way to get started is with our [managed Directus Cloud service](https://directus.cloud). 
+
+Create a new collection called `Messages`, with `date_created` and `user_created` fields enabled in the *Optional System Fields* pane on collection creation. Create a text field called `text`.
```
```suggestion
Create a new collection called `messages`, with `date_created` and `user_created` fields enabled in the *Optional System Fields* pane on collection creation. Create a text field called `text`.
```

paescuj replied:
Align collection name with the one used in subscription

paescuj committed:
Fix collection name to match name used in subscription

paescuj commented on docs/guides/real-time/chat/vue.md:
```diff
@@ -0,0 +1,399 @@
+
+You will need a Directus project. If you don’t already have one, the easiest way to get started is with our [managed Directus Cloud service](https://directus.cloud). 
+
+Create a new collection called `Messages`, with `date_created` and `user_created` fields enabled in the *Optional System Fields* pane on collection creation. Create a text field called `text`.
```
```suggestion
Create a new collection called `messages`, with `date_created` and `user_created` fields enabled in the *Optional System Fields* pane on collection creation. Create a text field called `text`.
```

paescuj commented on docs/guides/real-time/chat/react.md:
```diff
@@ -0,0 +1,422 @@
+You will need a Directus project. If you don’t already have one, the easiest way to get started is with our
+[managed Directus Cloud service](https://directus.cloud).
+
+Create a new collection called `Messages`, with `date_created` and `user_created` fields enabled in the _Optional System
```
```suggestion
Create a new collection called `messages`, with `date_created` and `user_created` fields enabled in the _Optional System
```

paescuj committed:
Fix collection name in other guides

paescuj commented on docs/guides/real-time/getting-started/graphql.md:
```diff
@@ -0,0 +1,91 @@
+
+## Create a Subscription
+
+After subscribing to collections over your connection, you will real-data changes in your collection. To subscribe to a
```
```suggestion
After subscribing to collections over your connection, you will receive real-time data changes in your collection. To subscribe to a
```

paescuj committed:
Fix diffs in doc & enhance chart example

paescuj committed:
Complete sentence in GraphQL guide

paescuj commented on docs/self-hosted/config-options.md:
```diff
@@ -1062,3 +1062,32 @@ Allows you to configure hard technical limits, to prevent abuse and optimize for
+| Variable           | Description                                | Default Value |
+|--------------------|--------------------------------------------|---------------|
+| `WEBSOCKETS_REST_ENABLED` | Whether or not to enable the REST message handlers. | `true` |
+| `WEBSOCKETS_REST_PATH` | The URL Path at which the WebSocket will be available. | `/websocket` |
```
```suggestion
| `WEBSOCKETS_REST_PATH` | The URL path at which the WebSocket endpoint will be available. | `/websocket` |
```

paescuj commented on docs/self-hosted/config-options.md:
```diff
@@ -1062,3 +1062,32 @@ Allows you to configure hard technical limits, to prevent abuse and optimize for
+| Variable           | Description                                | Default Value |
+|--------------------|--------------------------------------------|---------------|
+| `WEBSOCKETS_GRAPHQL_ENABLED` | Whether or not to enable the GraphQL Subscriptions. | `true` |
+| `WEBSOCKETS_GRAPHQL_PATH` | The URL Path at which the WebSocket will be available. | `/graphql` |
```
```suggestion
| `WEBSOCKETS_GRAPHQL_PATH` | The URL path at which the WebSocket endpoint will be available. | `/graphql` |
```

paescuj committed:
Small update to config description

paescuj reviewed (approved):
![giphy](https://github.com/directus/directus/assets/5363448/749aa9cd-1918-4f91-a1db-e9346ff1f615)
